quint/src/runtime/impl/compilerImpl.ts

```diff
@@ -1122,94 +1040,38 @@ export class CompilerVisitor implements IRVisitor {
    */
   private mapLambdaThenReduce(
     sourceId: bigint,
-    mapResultAndElems: (_array: Array<[RuntimeValue, RuntimeValue]>) => RuntimeValue
+    reduceFunction: (_array: Array<[RuntimeValue, RuntimeValue]>) => RuntimeValue
   ): void {
-    if (this.compStack.length < 2) {
-      this.errorTracker.addCompileError(sourceId, 'QNT501', 'Not enough arguments')
-      return
-    }
-    // lambda translated to Callable
-    const callable = this.compStack.pop() as Callable
-    // apply the lambda to a single element of the set
-    const evaluateElem = function (elem: RuntimeValue): Maybe<[RuntimeValue, RuntimeValue]> {
-      // evaluate the predicate against the actual arguments
-      const result = callable.eval([just(elem)])
-      return result.map(result => [result as RuntimeValue, elem])
-    }
-    this.applyFun(sourceId, 1, (iterable: Iterable<RuntimeValue>): Maybe<RuntimeValue> => {
-      return flatMap(iterable, evaluateElem).map(rs => mapResultAndElems(rs))
-    })
+    this.popArgs(2)
```

Comment:
starting here, I extracted some methods into functions defined at `operatorEvaluator.ts`. So these methods right now only deal with the stateful part, while the functions are responsible for the (more isolated) logic.